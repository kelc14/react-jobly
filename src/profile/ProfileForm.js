import React, { useState, useContext } from "react";
import UserContext from "../hooks/UserContext";
import "./Profile.css";
import JoblyApi from "../api/api";
import useAlerts from "../hooks/useAlerts";
import useErrors from "../hooks/useErrors";

/** ProfileForm component
 *
 * Updates user profile:
 * - first name
 * - last name
 * - email
 *
 *    => updates State [user] in app
 */

const ProfileForm = ({ updateUser }) => {
  const user = useContext(UserContext);
  let { username, firstName, lastName, email } = user;

  const INITIAL_STATE = { username, firstName, lastName, email };
  const [formData, setFormData] = useState(INITIAL_STATE);

  const [alert, setAlert, showAlerts] = useAlerts();
  const [err, documentErrors, showFormError] = useErrors();

  /** Send {firstName, lastName, password, email}
   * to API to patch user data
   *    & clear form. */

  const handleSubmit = async (evt) => {
    evt.preventDefault();
    delete formData.username;
    try {
      let res = await JoblyApi.updateUserDetails(formData, username);
      updateUser(res.user);
      setAlert([
        {
          message: "You successfully updated your profile.",
          type: "success",
        },
      ]);
    } catch (err) {
      documentErrors(err);
    }
  };

  /** Update local state w/curr state of input elem */

  const handleChange = (evt) => {
    const { name, value } = evt.target;
    setFormData((fData) => ({
      ...fData,
      [name]: value,
    }));
  };

  return (
    <div className="ProfileForm">
      <h2 className="ProfileForm-heading">PROFILE</h2>
      {alert.length > 0 && showAlerts()}
      <form className="ProfileForm-form" onSubmit={handleSubmit}>
        <label htmlFor="username" className="ProfileForm-label">
          Username:
        </label>
        <input
          id="username"
          type="text"
          name="username"
          onChange={handleChange}
          value={formData.username}
          className="ProfileForm-input"
          disabled
        />

        <label htmlFor="firstName" className="ProfileForm-label">
          First Name:
        </label>
        <input
          id="firstName"
          type="text"
          name="firstName"
          onChange={handleChange}
          value={formData.firstName}
          className="ProfileForm-input"
        />
        <label htmlFor="lastName" className="ProfileForm-label">
          Last Name:
        </label>
        <input
          id="lastName"
          type="text"
          name="lastName"
          onChange={handleChange}
          value={formData.lastName}
          className="ProfileForm-input"
        />
        <label htmlFor="email" className="ProfileForm-label">
          Email:
        </label>
        <input
          id="email"
          type="email"
          name="email"
          onChange={handleChange}
          value={formData.email}
          className="ProfileForm-input"
        />
        {err && showFormError()}

        <button className="ProfileForm-button">Update</button>
      </form>
    </div>
  );
};

export default ProfileForm;
