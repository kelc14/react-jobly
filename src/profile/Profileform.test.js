import React from "react";
import { render, screen, waitFor, act } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { MemoryRouter } from "react-router-dom";
import ProfileForm from "./ProfileForm";
import JoblyApi from "../api/api";
import UserContext from "../hooks/UserContext";

const customRender = (ui, { providerProps, ...renderOptions }) => {
  return render(
    <UserContext.Provider value={providerProps}>{ui}</UserContext.Provider>,
    renderOptions
  );
};

let providerProps;
beforeEach(
  () =>
    (providerProps = {
      username: "testingUser",
      firstName: "testFirst",
      lastName: "testLast",
      email: "email@email.com",
      isAdmin: false,
      jobs: [null],
    })
);
/**
 * SMOKE TEST
 */
test("ProfileForm renders without crashing", () => {
  customRender(<ProfileForm />, { providerProps });
});

/**  SNAPSHOT TEST */
it("renders without crashing", function () {
  const { asFragment } = customRender(<ProfileForm />, { providerProps });

  expect(asFragment()).toMatchSnapshot();
});

/** SIGNUP FORM e2e */
// works :)

it("logs a user in", async () => {
  const updateUser = jest.fn();
  customRender(<ProfileForm updateUser={updateUser} />, { providerProps });

  const usernameInput = screen.getByLabelText("Username:");
  const firstNameInput = screen.getByLabelText("First Name:");
  const lastNameInput = screen.getByLabelText("Last Name:");
  const emailInput = screen.getByLabelText("Email:");
  const updateBtn = screen.getByText("Update");

  // wrap in act when changing state
  act(() => {
    // userEvent.type(usernameInput, "testuser");
    userEvent.type(firstNameInput, "Test First1");
    userEvent.type(lastNameInput, "Test Last1");
    userEvent.type(emailInput, "email1@email.com");
  });

  // Mock the userSignup function
  const userUpdateMock = jest.fn().mockResolvedValue({
    user: {
      username: "testingUser",
      firstName: "Test First1",
      lastName: "Test Last1",
      email: "email1@email.com",
      isAdmin: false,
    },
  });
  jest.spyOn(JoblyApi, "updateUserDetails").mockImplementation(userUpdateMock);

  act(() => userEvent.click(updateBtn));

  // username and token (any string) returned
  await waitFor(() => {
    expect(updateUser).toHaveBeenCalledWith({
      username: "testingUser",
      firstName: "Test First1",
      lastName: "Test Last1",
      email: "email1@email.com",
      isAdmin: false,
    });
  });
});
