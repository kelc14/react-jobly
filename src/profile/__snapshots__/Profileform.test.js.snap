// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<DocumentFragment>
  <div
    class="ProfileForm"
  >
    <h2
      class="ProfileForm-heading"
    >
      PROFILE
    </h2>
    <form
      class="ProfileForm-form"
    >
      <label
        class="ProfileForm-label"
        for="username"
      >
        Username:
      </label>
      <input
        class="ProfileForm-input"
        disabled=""
        id="username"
        name="username"
        type="text"
        value="testingUser"
      />
      <label
        class="ProfileForm-label"
        for="firstName"
      >
        First Name:
      </label>
      <input
        class="ProfileForm-input"
        id="firstName"
        name="firstName"
        type="text"
        value="testFirst"
      />
      <label
        class="ProfileForm-label"
        for="lastName"
      >
        Last Name:
      </label>
      <input
        class="ProfileForm-input"
        id="lastName"
        name="lastName"
        type="text"
        value="testLast"
      />
      <label
        class="ProfileForm-label"
        for="email"
      >
        Email:
      </label>
      <input
        class="ProfileForm-input"
        id="email"
        name="email"
        type="email"
        value="email@email.com"
      />
      <button
        class="ProfileForm-button"
      >
        Update
      </button>
    </form>
  </div>
</DocumentFragment>
`;
